Add initial sphinx-gallery structure

djhoese opened the pull request:
This adds the necessary components to the docs to generate and include a gallery of examples using sphinx-gallery.

Getting help from the SciPy sprint (CC @SarthakJariwala).

TODO:

- [x] Generate image from example execution
- [x] Automatically close example after screenshot is generated
- [x] Update all example docstrings for better titles, tooltips, and generate descriptions
- [ ] Remove "# vispy: gallery N" comments at the top of the modules since they shouldn't be needed any more
- [x] Move examples from current directories to new sphinx gallery directories (ex. `examples/gloo`, `examples/scene`, `examples/plotting`)
- [x] Verify that download util stdout is not shown in CI generated website. Since the CI won't have any of the test data, it will have to download it when it runs the examples and that will be included in the gallery page for stdout.

Broken examples:

- [ ] `examples/basics/scene/stereo.py`
- [ ] `examples/basics/scene/shared_context.py`

@vispy/core Any questions so far?

![image](https://user-images.githubusercontent.com/1828519/126046549-093f8d64-51e1-4947-8ab0-1e464792aeec.png)

djhoese committed:
Add initial sphinx-gallery structure

djhoese committed:
Create initial sphinx-gallery custom scraper

djhoese committed:
Add gallery directory to .gitignore

djhoese committed:
Refactor gallery frame grabbing to be cleaner

djhoese commented:
It's happening...

![image](https://user-images.githubusercontent.com/1828519/126082710-b4627eed-21ae-4d97-9145-e8fe8633cb71.png)

djhoese committed:
More FrameGrabber refactor

djhoese committed:
More FrameGrabber refactoring

djhoese committed:
Fix gallery handling of multiple frames

djhoese commented:
In my last commits I got gif generation working. However, there is some example that is causing a segmentation fault and I still haven't figured out which one. Worst case it depends on the order examples are run.

larsoner commented on conf.py:
```diff
@@ -45,6 +48,148 @@
+        '../examples/plotting',
+        ]),
+    'only_warn_on_example_error': True,
+    'image_scrapers': (VisPyGalleryScraper(),),
```
A cleaner way to do this is to use the API contract here:

https://sphinx-gallery.github.io/stable/advanced.html#integrate-custom-scrapers-with-sphinx-gallery

You can see how PyVista does it here for example

https://github.com/pyvista/pyvista/blob/main/pyvista/__init__.py#L21

It both keeps `conf.py` more readable, and makes it so that any downstream libraries that want to use VisPy + SG can just do `image_scrapers = ('vispy',)` and it will magically work

djhoese replied:
Done! Thanks. Much cleaner.

larsoner replied:
Also has the nice side effect that now you can more easily add some basic unit tests for it if you want to be complete and avoid a code coverage drop :)

djhoese replied:
Tests added :wink:

larsoner commented on conf.py:
```diff
@@ -45,6 +48,148 @@
+        return self.__class__.__name__
+
+    def __call__(self, block, block_vars, gallery_conf):
+        """Scrape Matplotlib images.
```
update

larsoner reviewed (commented):
Looks like a good start, awesome that it seems to almost be working!

To prevent segfaults due to rapid close/reopen you might have to do the `_vispy_warmup` stuff we do in tests that add ~250 ms or so to each close/reopen cycle. For debugging these issues I'd also set up a `PATTERN=something make html-pattern` in the Makefile like in MNE:

https://github.com/mne-tools/mne-python/blob/main/doc/Makefile#L59-L62

djhoese committed:
Update sphinx-gallery to handle basic Qt applications

djhoese committed:
Add link from contributor's guide to source installation

djhoese committed:
Fix docstring in sphinx-gallery scraper

djhoese commented:
Good idea on the makefile pattern, I'll try that out. I'll see if I can get the scraper moved over too.

I've updated the scraper so it can naively handle larger Qt applications. The segfault seems to be caused by the example making its own QApplication instance and that interfering with other examples having already made the example. I've updated it so the example only uses the vispy app, but I'll need to document that it is Ok to use a regular QApplication (since vispy re-uses any existing one), but that examples aren't setup that way due to how they are executed when generating the gallery.

![image](https://user-images.githubusercontent.com/1828519/126194426-f0615a95-67b6-4e44-95d1-496464d1f6ca.png)

djhoese committed:
Fix bad linking in contrib guide

djhoese committed:
Add make html_dev-pattern helper target

djhoese committed:
Move scraper to vispy package

djhoese committed:
Add dev guide information for writing examples

djhoese committed:
Update writing examples with info on including generated images

djhoese committed:
Remove limit on example scripts to run for gallery

djhoese committed:
Remove sphinx-gallery reset modules

djhoese committed:
Split gallery into gloo, scene, plotting galleries

djhoese committed:
Update the gallery landing page to be more descriptive and helpful

djhoese committed:
Sort gallery examples by filename

djhoese committed:
Update gloo getting started to point to gloo gallery

djhoese committed:
Add example writing information about default of not grabbing frames

djhoese committed:
Add ignore for 'export.py' which has the wrong output

djhoese commented:
As an update to the TODO list at the top, I actually need the `# vispy: gallery N` comment to have flexiblity on what frame is captured as well as doing fancy stuff like getting multiple frames for animations.
